Add tests to check service config propagation

Ebaneck opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->
'tests', 'kubernetes', 'salt'

**Context**: 

See #2292 

**Summary**:

- Add tests to ensure that Service Configurations are propagated to the underlying services.

**Acceptance criteria**: 


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2292

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

Ebaneck committed:
tests: Add test to check propagation of service configurations

We need to be able to check that newly added service configurations
are propagated down the underlying services involved.

For this to happen, we have added tests to edit and apply a service
configuration then validate that the service runs with the newly
added configuration automatically

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2292-add-test-for-service-config-propagation` into
`development/2.5`
* [w/2.6/feature/2292-add-test-for-service-config-propagation](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/feature/2292-add-test-for-service-config-propagation) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

bert-e commented:
# History mismatch



Merge commit #e852ac2d6f639e6d7c13fa31a5d84601fff0d75a on the integration branch
`w/2.6/feature/2292-add-test-for-service-config-propagation` is merging a branch which is neither the current
branch `feature/2292-add-test-for-service-config-propagation` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `feature/2292-add-test-for-service-config-propagation` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented on tests/post/features/service_configuration.feature:
```diff
@@ -0,0 +1,11 @@
+    Scenario: Propagation of Service Configurations to underlying Services
+        Given the Kubernetes API is available
+        And pods with label 'app.kubernetes.io/name=dex' are 'Ready'
+        And we have 2 running pod labeled 'app.kubernetes.io/name=dex' in namespace 'metalk8s-auth' on node 'bootstrap'
```
Do not specify the `on node 'bootstrap'` please
```suggestion
        And we have 2 running pod labeled 'app.kubernetes.io/name=dex' in namespace 'metalk8s-auth'
```

TeddyAndrieux commented on tests/post/features/service_configuration.feature:
```diff
@@ -0,0 +1,11 @@
+        Given the Kubernetes API is available
+        And pods with label 'app.kubernetes.io/name=dex' are 'Ready'
+        And we have 2 running pod labeled 'app.kubernetes.io/name=dex' in namespace 'metalk8s-auth' on node 'bootstrap'
+        And we have a 'metalk8s-dex-config' ConfigMap with replicas count '2'
```
I think you should use something more generic than that, also here you match directly in the config and `ConfigMap` is not really a keyword here. 

IMO we should have something more generic that you can use for different test something like that
```
we have a 'metalk8s-dex-config' CSC in namespace 'metalk8s-auth' with '.spec.deployment.replicas' equal to '2'
```

So a fixture look like this 
```
we have a '<name>' CSC in namespace '<namespace>' with '<path>' equal to '<value>'
```

TeddyAndrieux commented on tests/post/features/service_configuration.feature:
```diff
@@ -0,0 +1,11 @@
+        And we have 2 running pod labeled 'app.kubernetes.io/name=dex' in namespace 'metalk8s-auth' on node 'bootstrap'
+        And we have a 'metalk8s-dex-config' ConfigMap with replicas count '2'
+        When I update the replicas count to '3'
+        And I apply a salt-state to propagate the changes to Dex Service
```
Those two should be more generic also

```
I update '<name>' CSC in namespace '<namespace>' '<path>' to '<value>'
```

```
I apply the '<salt states>' salt state
```

TeddyAndrieux replied:
Oh and btw to be consistent lets use `we` instead of `I`

TeddyAndrieux commented on tests/post/features/service_configuration.feature:
```diff
@@ -0,0 +1,11 @@
+        And we have a 'metalk8s-dex-config' ConfigMap with replicas count '2'
+        When I update the replicas count to '3'
+        And I apply a salt-state to propagate the changes to Dex Service
+        Then we have '3' available Deployments for the Dex app
```
```
we have 3 running pod labeled 'app.kubernetes.io/name=dex' in namespace 'metalk8s-auth'
```

Ebaneck replied:
Please see : https://github.com/scality/metalk8s/issues/2292#issuecomment-596614157

TeddyAndrieux replied:
Ok as you want but still need something more generic IMO like

```
we have '3' at '.spec.status.available_replicas' for 'dex' Deployment in namespace 'metalk8s-auth'
```

TeddyAndrieux reviewed (changes requested):
First quick review

TeddyAndrieux commented on tests/post/features/service_configuration.feature:
```diff
@@ -0,0 +1,10 @@
+        And pods with label 'app.kubernetes.io/name=dex' are 'Ready'
+        And we have 2 running pod labeled 'app.kubernetes.io/name=dex' in namespace 'metalk8s-auth'
+        And we have a 'metalk8s-dex-config' CSC in namespace 'metalk8s-auth' with '['spec']['deployment']['replicas']' equal to '2'
+        When we update 'metalk8s-dex-config' CSC in namespace 'metalk8s-auth' '['spec']['deployment']['replicas']' to '3'
```
I really prefer you use jsonpath syntax here (pretty easy in python you just have to split on `.` and get)

gdemonet replied:
You also have those `salt.utils` methods: [`salt.utils.data.traverse_dict`](https://github.com/saltstack/salt/blob/2018.3/salt/utils/data.py#L533) and [`salt.utils.data.traverse_dict_and_list`](https://github.com/saltstack/salt/blob/2018.3/salt/utils/data.py#L551)

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,261 @@
+            name, namespace, yaml.safe_load(BODY.format(
+                replicas_count=value)
+            )
+        )
```
No please do not override the full config,
Retrieve it load the yaml and edit the field you want and then push it back

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,261 @@
+@given(parsers.parse(
+    "we have a '{name}' CSC in namespace '{namespace}' with "
+    "'{path}' equal to '{value}'"))
+def check_configmap_configuration(
```
Minor: 
```suggestion
def check_csc_configuration(
```

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,261 @@
+
+    assert int(value) == response_value, (
+        "Expected value {} for key {}, got {}".format(
+            value, path, response_value
```
Please add the name of the config map and the namespace in the message

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,261 @@
+        raise Exception(
+            'Failed loading `config.yaml` from ConfigMap {}: {!s}'.format(
+                name, exc
+            )
```
Add the namespace also please

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,261 @@
+    except yaml.YAMLError as exc:
+        raise Exception(
+            'Invalid YAML format in ConfigMap {}: {!s}'.format(name, exc)
+        )
```
+ namespace :smiley:

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,261 @@
+        )
+    except yaml.YAMLError as exc:
+        raise Exception(
+            'Invalid YAML format in ConfigMap {}: {!s}'.format(name, exc)
```
ditto

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,261 @@
+        )
+    except Exception as exc:
+        raise Exception(
+            'Failed loading `config.yaml` from ConfigMap {}: {!s}'.format(
```
dittoto

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,261 @@
+    output = utils.run_salt_command(host, cmd, ssh_config)
+
+    assert output.exit_status == 0, \
+        'Failed applying salt-state: \nout: {}\nerr:'.format(
```
Add the name of the salt state you try to apply please

Ebaneck committed:
tests: Add a generic utility for running commands in a salt-master container

Applying salt-states to minions require that we execute commands in the
salt-master container.
This commit adds a generic utility to `utils.py` making it generic and
callable from the entire test suite

Closes: 2292

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,271 @@
+
+
+def traversed_dict(data, key):
+
```
Remove this empty line please

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,271 @@
+def traversed_dict(data, key):
+
+    '''
+    Traverse a dict using a 'delimiter' on a  target string.
```
Minor: SInce you talk about delimiter either you should add a delimiter arg to this function (that default to `.` or talk about dot here (personnally prefer to have delimiter as argument :) )

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,271 @@
+        k8s_client, name, namespace, patch
+    )
+
+    return response
```
Hmm why ? Do we really care about it ? (Also here `metalk8s-dex-config` is hardcoded not really like it)

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,271 @@
+    patch = \
+        {'data': {'config.yaml': yaml.safe_dump(
+            csc_obj, default_flow_style=False)
+        }}
```
Minor: Personnaly don't really like this indentation
```suggestion
    patch = {
        'data': {
            'config.yaml': yaml.safe_dump(csc_obj, default_flow_style=False)
        }
    }
```

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,271 @@
+                )
+            )
+
+        assert int(value) == eval("response"+"{}".format(path)), (
```
??? I think you forgot to use `traversed_dict` here.

And global python comment: better to avoid using `eval` as much as possible

slaperche-scality replied:
Yeah, `eval` is a no-go.

slaperche-scality commented on tests/utils.py:
```diff
@@ -104,3 +104,49 @@ def requests_retry_session(
+        namespace='kube-system'
+    )
+
+    return output
```
In `install/steps/test_expansion.py` we have an helper like that:

```python
def run_salt_command(host, command, ssh_config):
    """Run a command inside the salt-master container."""

    pod = 'salt-master-{}'.format(
        utils.get_node_name('bootstrap', ssh_config)
    )

    output = kubectl_exec(
        host,
        command,
        pod,
        container='salt-master',
        namespace='kube-system'
    )

    assert output.exit_status == 0, \
        'deploy failed with: \nout: {}\nerr:'.format(
            output.stdout,
            output.stderr
        )
```

You should either reuse it/move it here (with a bit of rework) OR replace it by yours.
But we shouldn't keep both.

Ebaneck replied:
ok

slaperche-scality commented on tests/utils.py:
```diff
@@ -104,3 +104,49 @@ def requests_retry_session(
     return session
+
+
+def kubectl_exec(
+    host,
+    command,
+    pod,
+    kubeconfig='/etc/kubernetes/admin.conf',
+    **kwargs
+):
+    """Grab the return code from a `kubectl exec`"""
+    kube_args = ['--kubeconfig', kubeconfig]
+
+    if kwargs.get('container'):
+        kube_args.extend(['-c', kwargs.get('container')])
+    if kwargs.get('namespace'):
+        kube_args.extend(['-n', kwargs.get('namespace')])
+
+    kubectl_cmd_tplt = 'kubectl exec {} {} -- {}'
+
+    with host.sudo():
+        output = host.run(
+            kubectl_cmd_tplt.format(
+                pod,
+                ' '.join(kube_args),
+                ' '.join(command)
+            )
+        )
+        return output
```
In `install/steps/test_expansion.py` we already have an helper like that:

```python
def kubectl_exec(
    host,
    command,
    pod,
    kubeconfig='/etc/kubernetes/admin.conf',
    **kwargs
):
    """Grab the return code from a `kubectl exec`"""
    kube_args = ['--kubeconfig', kubeconfig]

    if kwargs.get('container'):
        kube_args.extend(['-c', kwargs.get('container')])
    if kwargs.get('namespace'):
        kube_args.extend(['-n', kwargs.get('namespace')])

    kubectl_cmd_tplt = 'kubectl exec {} {} -- {}'

    with host.sudo():
        output = host.run(
            kubectl_cmd_tplt.format(
                pod,
                ' '.join(kube_args),
                ' '.join(command)
            )
        )
        return output
```

You should either reuse it/move it here (with a bit of rework) OR replace it by yours.
But we shouldn't keep both.

slaperche-scality commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,271 @@
+                exc
+            )
+        )
+    return ptr
```
I think this could be simplified as:

```python
def traversed_dict(data, key, delimiter='.'):
    return functools.reduce(operator.getitem, key.split(delimiter), data)
```

Ebaneck committed:
test: Refactor expansion tests to use kubectl exec utility helper

In the previous commit, we added the kubectl_exec helper that makes it
possible to execute a command in the salt-master container to the utils
section of the test.

In this commit, we refactor the `expansion tests` to reuse this utility
ensuring that we do not have duplicate code

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,229 @@
+            state,
+            output.stdout,
+            output.stderr
+        )
```
I think this assert could be directly in `run_salt_command`

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,229 @@
+    return response
+
+
+def yaml_load(full_csc, name, namespace):
```
Minor: Maybe just rename this function as it's not used to load a "yaml" but specific to CSC maybe "csc_load"

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,229 @@
+
+    full_csc = read_namespace_config_map(k8s_client, name, namespace)
+    csc_obj = yaml_load(full_csc, name, namespace)
+    csc_obj['spec']['deployment']['replicas'] = int(value)
```
Nop, use `path` also here

TeddyAndrieux replied:
I also don't really like the hardcoded `int` here because we want to use this same function for others service configuration (that may be not int).

Maybe we should have 2 separate `when` one with quoted `{path}` for string and one without for int

Then you should create an helper to update service configuration that take the value and do no cast it

slaperche-scality replied:
> Maybe we should have 2 separate `when` one with quoted `{path}` for string and one without for int

I think you can use a single `when`, and it can works fine if you use `"foo"` for string and `4` for integer.

`eval` is really bad, but [`literal_eval`](https://docs.python.org/3.6/library/ast.html#ast.literal_eval) ("Safely evaluate an expression node or a string containing a Python literal") is pretty cool.

```
>>> ast.literal_eval('3')
3
>>> ast.literal_eval('"foo"')
'foo'
```

TeddyAndrieux replied:
oh ok i didn't know this one, cool :+1:

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,239 @@
+        if isinstance(v, dict):
+            set_dict_element(v, path, new_value)
+        if k == required_key:
+            data[k] = new_value
```
What ? No !!

Use `.get` or like I suggest you, you can basically use `get_dict_element` with `path.split(delimiter)[:-1]` (with a join or making `get_dict_element` handle list (which is the cleaner solution IMO, except if we want to properly handle KeyError, but I think we don't really care here))

slaperche-scality replied:
Instead of making `get_dict_element` handling both a list and a string (which is not *that* clean) I think it can also be expressed as:

```python
def set_dict_element(data, path, value, delimiter='.'):
    path, _, key = path.rpartition('.')
    (get_dict_element(data, path) if path else data)[key] = value
    return data
```

Ebaneck replied:
ok

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,239 @@
+            set_dict_element(v, path, new_value)
+        if k == required_key:
+            data[k] = new_value
+    return data
```
Minor: Btw this return is not mandatory, it's a dict so anyway the input "data" will change

TeddyAndrieux commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -0,0 +1,239 @@
+    return functools.reduce(operator.getitem, path.split(delimiter), data)
+
+
+def set_dict_element(data, path, new_value, delimiter='.'):
```
I think `value` is enough do not need to specify `new_`

TeddyAndrieux replied:
Minor: Maybe you could move this two function in `utils` as we may want to use them somewhere else

Ebaneck committed:
tests: Add utility for setting and getting values from a nested dictionary

To patch ConfigMaps, we need to be able to get and set new values
from the ConfigMap object returned by the Kubernetes API.

For this to happen, this commit adds two utilities to `utils.py`
that ensure we can get and set values in a classic python dictionary
given that key path is string written in the format `'spec.status.replicas'`

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2292-add-test-for-service-config-propagation` into
`development/2.5`
* [w/2.6/feature/2292-add-test-for-service-config-propagation](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/feature/2292-add-test-for-service-config-propagation) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

bert-e commented:
# History mismatch



Merge commit #b866f4118e37f064ed86822f2366fe0c5ccff880 on the integration branch
`w/2.6/feature/2292-add-test-for-service-config-propagation` is merging a branch which is neither the current
branch `feature/2292-add-test-for-service-config-propagation` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `feature/2292-add-test-for-service-config-propagation` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux reviewed (dismissed):
LGTM

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2292-add-test-for-service-config-propagation` into
`development/2.5`
* [w/2.6/feature/2292-add-test-for-service-config-propagation](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/feature/2292-add-test-for-service-config-propagation) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

Ebaneck commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/b3889aecf6abd8de694a773c99bb784b81321a6a) did not succeed in branch feature/2292-add-test-for-service-config-propagation.




*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #322c4ccaceb23f1b89131759659f6338c90cde38 on the integration branch
`w/2.6/feature/2292-add-test-for-service-config-propagation` is merging a branch which is neither the current
branch `feature/2292-add-test-for-service-config-propagation` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `feature/2292-add-test-for-service-config-propagation` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2292-add-test-for-service-config-propagation` into
`development/2.5`
* [w/2.6/feature/2292-add-test-for-service-config-propagation](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/feature/2292-add-test-for-service-config-propagation) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort










*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/f66a52b81563507535af92a5f6c5df75b6a9e3e9) did not succeed in branch feature/2292-add-test-for-service-config-propagation.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/4c5362f9b642c9a2f0d2d462f72ce31e901d280f) did not succeed in branch w/2.6/feature/2292-add-test-for-service-config-propagation.




*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #ef63822e6bc2ccb7cd8299b5bb462b7ca1bc7a28 on the integration branch
`w/2.6/feature/2292-add-test-for-service-config-propagation` is merging a branch which is neither the current
branch `feature/2292-add-test-for-service-config-propagation` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `feature/2292-add-test-for-service-config-propagation` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2292-add-test-for-service-config-propagation` into
`development/2.5`
* [w/2.6/feature/2292-add-test-for-service-config-propagation](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/feature/2292-add-test-for-service-config-propagation) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



Please check the status of the associated issue None.

Goodbye ebaneck.
